File: frontend/lib/models/project.dart
class Project {
  final String id;
  final String slug;
  final String name;
  final String description;
  final String groupId;
  final String createdBy;
  final String? creatorUsername;
  final DateTime createdAt;
  final String? presenterId;
  final String? presenterUsername;
  final bool enableProblem;
  final bool enableVote;
  final bool enablePrioritise;
  final bool guestsEnabled;
  final String problemStatement;
  final int activeUsers;
  final String interstitialProblem;
  final String interstitialBrainstorm;
  final String interstitialGroup;
  final String interstitialVote;
  final String interstitialPrioritise;

  const Project({
    required this.id,
    required this.slug,
    required this.name,
    required this.description,
    required this.groupId,
    required this.createdBy,
    this.creatorUsername,
    required this.createdAt,
    this.presenterId,
    this.presenterUsername,
    this.enableProblem = true,
    this.enableVote = true,
    this.enablePrioritise = true,
    this.guestsEnabled = false,
    this.problemStatement = '',
    this.activeUsers = 0,
    this.interstitialProblem = '',
    this.interstitialBrainstorm = '',
    this.interstitialGroup = '',
    this.interstitialVote = '',
    this.interstitialPrioritise = '',
  });

  /// Returns the custom interstitial description for [slide], or null to use
  /// the app default.
  String? interstitialForSlide(int slide) {
    final value = switch (slide) {
      0 => interstitialProblem,
      1 => interstitialBrainstorm,
      2 => interstitialGroup,
      3 => interstitialVote,
      4 => interstitialPrioritise,
      _ => '',
    };
    return value.isEmpty ? null : value;
  }

  factory Project.fromJson(Map<String, dynamic> json) => Project(
        id: json['id'] as String,
        slug: json['slug'] as String,
        name: json['name'] as String,
        description: json['description'] as String? ?? '',
        groupId: json['group_id'] as String,
        createdBy: json['created_by'] as String,
        creatorUsername:
            (json['creator'] as Map<String, dynamic>?)?['username'] as String?,
        createdAt: DateTime.parse(json['created_at'] as String),
        presenterId: json['presenter_id'] as String?,
        presenterUsername:
            (json['presenter'] as Map<String, dynamic>?)?['username'] as String?,
        enableProblem: json['enable_problem'] as bool? ?? true,
        enableVote: json['enable_vote'] as bool? ?? true,
        enablePrioritise: json['enable_prioritise'] as bool? ?? true,
        guestsEnabled: json['guests_enabled'] as bool? ?? false,
        problemStatement: json['problem_statement'] as String? ?? '',
        activeUsers: json['active_users'] as int? ?? 0,
        interstitialProblem: json['interstitial_problem'] as String? ?? '',
        interstitialBrainstorm: json['interstitial_brainstorm'] as String? ?? '',
        interstitialGroup: json['interstitial_group'] as String? ?? '',
        interstitialVote: json['interstitial_vote'] as String? ?? '',
        interstitialPrioritise: json['interstitial_prioritise'] as String? ?? '',
      );

  Project copyWith({String? presenterId, String? presenterUsername, bool clearPresenter = false}) =>
      Project(
        id: id,
        slug: slug,
        name: name,
        description: description,
        groupId: groupId,
        createdBy: createdBy,
        creatorUsername: creatorUsername,
        createdAt: createdAt,
        presenterId: clearPresenter ? null : (presenterId ?? this.presenterId),
        presenterUsername: clearPresenter ? null : (presenterUsername ?? this.presenterUsername),
        enableProblem: enableProblem,
        enableVote: enableVote,
        enablePrioritise: enablePrioritise,
      );
}

File: backend/models/project.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Project struct {
	ID          string         `json:"id"          gorm:"primaryKey;type:text"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `json:"-"           gorm:"index"`
	Name        string         `json:"name"        gorm:"not null"`
	Slug        string         `json:"slug"        gorm:"uniqueIndex:idx_group_project_slug;not null;type:text"`
	Description string         `json:"description"`
	GroupID     string         `json:"group_id"    gorm:"not null;index;uniqueIndex:idx_group_project_slug;type:text"`
	CreatedBy   string         `json:"created_by"  gorm:"type:text"`
	Creator     *User          `json:"creator,omitempty"   gorm:"foreignKey:CreatedBy"`
	PresenterID      *string        `json:"presenter_id"        gorm:"type:text"`
	Presenter        *User          `json:"presenter,omitempty" gorm:"foreignKey:PresenterID"`
	EnableProblem      bool   `json:"enable_problem"      gorm:"not null;default:true"`
	EnableVote         bool   `json:"enable_vote"         gorm:"not null;default:true"`
	EnablePrioritise   bool   `json:"enable_prioritise"   gorm:"not null;default:true"`
	ProblemStatement   string `json:"problem_statement"   gorm:"type:text;not null;default:''"`

	GuestsEnabled bool   `json:"guests_enabled" gorm:"not null;default:false"`

	// Custom interstitial descriptions (empty = use app default)
	InterstitialProblem    string `json:"interstitial_problem"    gorm:"type:text;not null;default:''"`
	InterstitialBrainstorm string `json:"interstitial_brainstorm" gorm:"type:text;not null;default:''"`
	InterstitialGroup      string `json:"interstitial_group"      gorm:"type:text;not null;default:''"`
	InterstitialVote       string `json:"interstitial_vote"       gorm:"type:text;not null;default:''"`
	InterstitialPrioritise string `json:"interstitial_prioritise" gorm:"type:text;not null;default:''"`
}

func (p *Project) BeforeCreate(tx *gorm.DB) error {
	if p.ID == "" {
		p.ID = uuid.New().String()
	}
	return nil
}
